snapcommunicator/doc/man/snapclusterstatus.1
.TH "snapclusterstatus" 1 "Sat Apr 20 2019" "Version 1.7.7" "snapcommunicator" \" -*- nroff -*-
.ad l
.nh
.SH NAME
snapclusterstatus - query snapcommunicator about the cluster status
.SH SYNOPSIS
.B snapclusterstatus
[\fB--help\fR] [\fB--version\fR] [\fB\-c|\-\-config\fR \fI\,config\/\fR]
.SH DESCRIPTION

The \fBsnapclusterstatus\fR sends a \fBCLUSTERSTATUS\fR message to the
\fBsnapcommunicator\fR daemon, prints out the results, and exits.

If the \fBsnapcommunicator\fR daemon is not running, then an error
occurs and corresponding messages are printed.

By default the \fBsnapclusterstatus\fR uses the configuration information
found in \fB/etc/snapwebsites/snapcommunicator.conf\fR and 
\fB/etc/snapwebsites/snapwebsites.d/snapcommunicator.conf\fR. The filename
can be used using the \fB--config\fR option.

The command supports the following options:

.TP
\fB\-\-help
prints out the help page

.TP
\fB\-\-version
prints out the version (which is really snapcommunicator's version)

.TP
\fB\-\-config | \-c \fI\,config\/\fR
specify a different configuration file than the default.

.SH AUTHOR

Written by Alexis Wilke, based on snaplock.

.SH REPORTING BUGS

https://github.com/m2osw/snapwebsites/issues

.SH COPYRIGHT

Copyright (c) 2019  Made to Order Software Corp.
License GPLv2+.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

.SH SEE ALSO
Full documentation at: <http://snapwebsites.org/>


debian/changelog
snapwebsites (1.7.9.0~xenial) xenial; urgency=high

  * Added support for a pool of threads and workers.

 -- Alexis Wilke <alexis@m2osw.com>  Tue, 23 Apr 2019 23:36:44 -0700

snapwebsites (1.7.8.1~xenial) xenial; urgency=high

  * Bump version to run a build against the new cassandra library.

 -- Alexis Wilke <alexis@m2osw.com>  Tue, 23 Apr 2019 00:58:44 -0700

snapwebsites (1.7.8.0~xenial) xenial; urgency=high

  * SNAP-684: Added the snapclusterstatus command line tool.
  * SNAP-684: Included a man.1 page for snapclusterstatus.
  * SNAP-684: Made snapcommunicator save the snap-cluster.txt file with its
    most current status.
  * SNAP-684: Fixed a bug in a one computer environment the cluster status
    was not set until we receive a very first CLUSTERSTATUS message.
  * EX-204: Fixed the path used to check the field, was using wrong variable.
  * EX-204: Made the upload reset button appear only if a flag is turned on.
  * In the snapmanagercgi output, fixed the page title so it includes the
    host name.
  * Fixed the double "(Select Host)" message from the status.

 -- Alexis Wilke <alexis@m2osw.com>  Sun, 21 Apr 2019 00:14:19 -0700

snapwebsites (1.7.7.0~xenial) xenial; urgency=high

  * SNAP-181: Applied a fix as the whitelist would not update automatically,
    we have to save the value in the config (because it's static memory.)
  * SNAP-181: Fixed the `signal_secret` which would also not get updated.
  * SNAP-350: Fixed list of supported message in snaplock, the CLUSTERUP
    and CLUSTERDOWN had to be included.
  * SNAP-478: Fixed the type for result for a DELETE in the index plugin.
  * SNAP-478: Fixed the index case, when indexing the key may change, to
    avoid duplicates, we must delete existing entries that do not match
    the new 'column1' value.
  * Fixed the menu, do not show Host Watchdog when no host is selected.

 -- Alexis Wilke <alexis@m2osw.com>  Fri, 19 Apr 2019 20:30:20 -0700

snapwebsites (1.7.6.0~xenial) xenial; urgency=high

  * SNAP-181: added a field to edit the "whitelist" variable from iplock.

 -- Alexis Wilke <alexis@m2osw.com>  Wed, 17 Apr 2019 00:00:20 -0700

snapwebsites (1.7.5.0~xenial) xenial; urgency=high

  * SNAP-526: added a field in snapmanager to edit the `signal_secret` field.
  * SNAP-478: added the index plugin to the packages.

 -- Alexis Wilke <alexis@m2osw.com>  Mon, 15 Apr 2019 21:56:32 -0700

snapwebsites (1.7.4.0~xenial) xenial; urgency=high

  * SNAP-413 and SNAP-421 and SNAP-454: Moved the snapcreatecontext
    functionality to snapdbproxy.
  * SNAP-413: Moved the snapcreatetables functionality to snapdbproxy.
  * SNAP-413: Fixed the NEWTABLE implementation.
  * SNAP-267: Added an XSD to verify tables.xml files and updated the
    CMakeLists.txt.
  * SNAP-478: implemented the index plugin to have instant "lists".
  * SNAP-478: added a link() function in snap_expr, useful for indexes.
  * SNAP-679: added a "temporary" fix to the fail2ban nice value.
  * SNAP-492: implemented the support for a PID file for snapserver,
    snabackend, snacommunicator, and snapwatchdog (so far.)
  * SNAP-641: Started work on a fix to not leak messages to the wrong user.
  * SNAP-526: added the ability to add a secret code in a UDP message to
    make them a little more secure.
  * Changed the sendmail algorithm to allow for ignoring the fact that the
    user's email address failed with a bounce before if the user is an admin.
  * When marking a session as used up, save the date when the event happened.
  * Install the unblock-ip script even though it's not very useful as it stands.
  * Converted the snapcommunicator address test to make use of addr::addr
    (the domain name includes the port!)
  * Added a script one can use to search for users who have a bounced email.
  * Add the address to the addrinfo_t unique_ptr<>() because it could be that
    we had an error (r != 0) and a pointer (addrinfo != null).
  * Verify buffer length in get_client_addr().
  * Added support for keepalive in the client BIO.
  * Properly initialize a variable in case BIO_get_fd() fails.
  * Applied a fix to the snapmanager so it does not crash if the name of the
    host specified on the query string is unknown.
  * Fixed the cassandra plugin, if the "listen_address" is not defined it
    would _crash_ when trying to get the parameter when it's not defined.
  * Applied a fix to the snapstop making the `--service` parameter optional.
    (The `--service "$MAINPID"` becomes `--service` when $MAINPID is empty.)
  * Added /etc/vim/vimrc.local so we get our search, line numbers, etc.
  * Made the attach/detach session RAII so it at least happens on exceptions.
  * Added function to create a complete page in one call. (editor create_page)
  * Fixed the f_cassandra_connect_timer_index calculation.
  * Made the type found by find_type_with() available through another call.
  * Fixed the default watchdog plugins list, "package" was wrong, "s" missing.
  * Applied a fix to the message box, it adds a scrollbar when many messages
    appear in the list.
  * Fixed HUGE bug in content::modified_content_impl() where I would return
    false if the snap_child object was ready. That test was inverted!
  * Various clean ups.

 -- Alexis Wilke <alexis@m2osw.com>  Tue, 05 Mar 2019 10:49:22 -0700

snapwebsites (1.7.3.0~xenial) xenial; urgency=high

  * Remove reference to /run/snapwebsites from snapbuild.service.

 -- Alexis Wilke <alexis@m2osw.com>  Sat, 27 Oct 2018 11:37:48 -0700

snapwebsites (1.7.2.0~xenial) xenial; urgency=high

  * Make use of snapbase.tmpfile to create the /run/snapwebsites folder.

 -- Alexis Wilke <alexis@m2osw.com>  Thu, 25 Oct 2018 09:47:04 -0700

snapwebsites (1.7.1.0~xenial) xenial; urgency=high

  * Previous throwing when a plugin version is not correct.

 -- Alexis Wilke <alexis@m2osw.com>  Wed, 24 Oct 2018 01:34:18 -0700

snapwebsites (1.7.0.0~xenial) xenial; urgency=high

  * Create a new official version, this is much more stable than 1.6.

snapcommunicator/src/snapclusterstatus.cpp
// Snap Websites Server -- check the current status of the cluster
// Copyright (c) 2019  Made to Order Software Corp.  All Rights Reserved
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

// ourselves
//
#include "version.h"

// snapwebsites lib
//
#include <snapwebsites/snapwebsites.h>
#include <snapwebsites/snap_communicator.h>
#include <snapwebsites/snap_communicator_dispatcher.h>
#include <snapwebsites/snap_config.h>
#include <snapwebsites/not_reached.h>



namespace
{


class snapcluster;

class snapcluster_messenger
    : public snap::snap_communicator::snap_tcp_client_message_connection
{
public:
    typedef std::shared_ptr<snapcluster_messenger>     pointer_t;

                                snapcluster_messenger(snapcluster * sl, std::string const & addr, int port);
                                snapcluster_messenger(snapcluster_messenger const & rhs) = delete;
    virtual                     ~snapcluster_messenger() override {}

    snapcluster_messenger &     operator = (snapcluster_messenger const & rhs) = delete;

protected:
    // this is owned by a snaplock function so no need for a smart pointer
    // (and it would create a loop)
    snapcluster *               f_snapcluster = nullptr;
};


snapcluster_messenger::snapcluster_messenger(snapcluster * sl, std::string const & addr, int port)
    : snap_tcp_client_message_connection(addr, port)
    , f_snapcluster(sl)
{
    set_name("snapcluster messenger");
}



#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wnon-virtual-dtor"
class snapcluster
    : public snap::snap_communicator::connection_with_send_message
    , public snap::dispatcher<snapcluster>
    , public std::enable_shared_from_this<snapcluster>
{
public:
    typedef std::shared_ptr<snapcluster>      pointer_t;

                                snapcluster(int argc, char * argv[]);
                                snapcluster(snapcluster const & rhs) = delete;
    virtual                     ~snapcluster() {}

    snapcluster &               operator = (snapcluster const & rhs) = delete;

    void                        run();

    // implementation of connection_with_send_message
    virtual bool                send_message(snap::snap_communicator_message const & message, bool cache = false) override;

    // implementation of snap::snapcommunicator::connection_with_send_message
    virtual void                ready(snap::snap_communicator_message & message) override; // no "msg_" because that's in connection_with_send_message
    virtual void                stop(bool quitting) override; // no "msg_" because that's in connection_with_send_message

private:
    void                        usage(advgetopt::getopt::status_t status);
    void                        done(snap::snap_communicator_message & message);

    // messages handled by the dispatcher
    // (see also ready() and stop() above)
    //
    void                        msg_cluster_status(snap::snap_communicator_message & message);
    void                        msg_cluster_complete(snap::snap_communicator_message & message);

    static snap::dispatcher<snapcluster>::dispatcher_match::vector_t const
                                        g_snapcluster_service_messages;

    advgetopt::getopt                   f_opt;
    snap::snap_config                   f_config;
    QString                             f_communicator_addr = QString("localhost");
    int                                 f_communicator_port = 4040;
    snap::snap_communicator::pointer_t  f_communicator = snap::snap_communicator::pointer_t();
    snapcluster_messenger::pointer_t    f_messenger = snapcluster_messenger::pointer_t();
    QString                             f_cluster_status = QString();       // UP or DOWN
    QString                             f_cluster_complete = QString();     // COMPLETE or INCOMPLETE
    size_t                              f_neighbors_count = 0;
};
#pragma GCC diagnostic pop



/** \brief List of snapcluster commands.
 *
 * The following table defines the commands understood by snapcluster,
 * which are pretty limited, mainly we want to gather the status from
 * the snapcommunicator process.
 */
snap::dispatcher<snapcluster>::dispatcher_match::vector_t const snapcluster::g_snapcluster_service_messages =
{
    {
        "CLUSTERUP"
      , &snapcluster::msg_cluster_status
    },
    {
        "CLUSTERDOWN"
      , &snapcluster::msg_cluster_status
    },
    {
        "CLUSTERCOMPLETE"
      , &snapcluster::msg_cluster_complete
    },
    {
        "CLUSTERINCOMPLETE"
      , &snapcluster::msg_cluster_complete
    }
};


const std::vector<std::string> g_configuration_files; // Empty

advgetopt::getopt::option const g_snapcluster_options[] =
{
    {
        '\0',
        advgetopt::getopt::GETOPT_FLAG_SHOW_USAGE_ON_ERROR,
        nullptr,
        nullptr,
        "Usage: %p [-<opt>]",
        advgetopt::getopt::argument_mode_t::help_argument
    },
    {
        '\0',
        advgetopt::getopt::GETOPT_FLAG_SHOW_USAGE_ON_ERROR,
        nullptr,
        nullptr,
        "where -<opt> is one or more of:",
        advgetopt::getopt::argument_mode_t::help_argument
    },
    {
        'c',
        advgetopt::getopt::GETOPT_FLAG_ENVIRONMENT_VARIABLE | advgetopt::getopt::GETOPT_FLAG_SHOW_USAGE_ON_ERROR,
        "config",
        nullptr,
        "path to the snapcommunicator configuration file",
        advgetopt::getopt::argument_mode_t::optional_argument
    },
    {
        '\0',
        advgetopt::getopt::GETOPT_FLAG_SHOW_USAGE_ON_ERROR,
        "help",
        nullptr,
        "show this help output",
        advgetopt::getopt::argument_mode_t::no_argument
    },
    {
        '\0',
        advgetopt::getopt::GETOPT_FLAG_SHOW_USAGE_ON_ERROR,
        "version",
        nullptr,
        "show the version of %p and exit",
        advgetopt::getopt::argument_mode_t::no_argument
    },
    {
        '\0',
        0,
        nullptr,
        nullptr,
        nullptr,
        advgetopt::getopt::argument_mode_t::end_of_options
    }
};


snapcluster::snapcluster(int argc, char * argv[])
    : dispatcher(this, g_snapcluster_service_messages)
    , f_opt(argc, argv, g_snapcluster_options, g_configuration_files, nullptr)
    , f_config("snapcommunicator")
{
    // --help
    if(f_opt.is_defined("help"))
    {
        usage(advgetopt::getopt::status_t::no_error);
        snap::NOTREACHED();
    }

    // --version
    if(f_opt.is_defined("version"))
    {
        std::cerr << SNAPCOMMUNICATOR_VERSION_STRING << std::endl;
        exit(1);
        snap::NOTREACHED();
    }

    add_snap_communicator_commands();

    // read the configuration file
    //
    if(f_opt.is_defined("config"))
    {
        f_config.set_configuration_path(f_opt.get_string("config"));
    }

    tcp_client_server::get_addr_port(
                  QString::fromUtf8(f_config("snapcommunicator", "local_listen").c_str())
                , f_communicator_addr
                , f_communicator_port
                , "tcp");
}


void snapcluster::run()
{
    f_communicator = snap::snap_communicator::instance();

    f_messenger.reset(new snapcluster_messenger(this
                                              , f_communicator_addr.toUtf8().data()
                                              , f_communicator_port));
    f_messenger->set_dispatcher(shared_from_this());
    f_communicator->add_connection(f_messenger);

    // our messenger here is a direct connection (not a permanent one) so
    // we have to REGISTER immediately (if it couldn't connect we already
    // threw so this works)
    //
    snap::snap_communicator_message register_snapcluster;
    register_snapcluster.set_command("REGISTER");
    register_snapcluster.add_parameter("service", "snapcluster");
    register_snapcluster.add_parameter("version", snap::snap_communicator::VERSION);
    f_messenger->send_message(register_snapcluster);

    f_communicator->run();
}


bool snapcluster::send_message(snap::snap_communicator_message const & message, bool cache)
{
    return f_messenger->send_message(message, cache);
}


void snapcluster::ready(snap::snap_communicator_message & message)
{
    snap::NOTUSED(message);

    snap::snap_communicator_message clusterstatus_message;
    clusterstatus_message.set_command("CLUSTERSTATUS");
    clusterstatus_message.set_service("snapcommunicator");
    send_message(clusterstatus_message);
}


void snapcluster::stop(bool quitting)
{
    snap::NOTUSED(quitting);

    if(f_messenger != nullptr)
    {
        f_communicator->remove_connection(f_messenger);
        f_messenger.reset();
    }
}


void snapcluster::usage(advgetopt::getopt::status_t status)
{
    f_opt.usage(status, "snapcluster");
    exit(1);
}


void snapcluster::msg_cluster_status(snap::snap_communicator_message & message)
{
    f_cluster_status = message.get_command();
    done(message);
}


void snapcluster::msg_cluster_complete(snap::snap_communicator_message & message)
{
    f_cluster_complete = message.get_command();
    done(message);
}


void snapcluster::done(snap::snap_communicator_message & message)
{
    if(f_cluster_status.isEmpty()
    || f_cluster_complete.isEmpty())
    {
        // not quite done yet...
        return;
    }

    f_neighbors_count = message.get_integer_parameter("neighbors_count");

    // got our info!
    //
    std::cout << "              Status: " << f_cluster_status          << std::endl
              << "            Complete: " << f_cluster_complete        << std::endl
              << "Computers in Cluster: " << f_neighbors_count         << std::endl
              << " Quorum of Computers: " << f_neighbors_count / 2 + 1 << std::endl;

    // we're done, remove the messenger which is enough for the
    // snap_communicator::run() to return
    //
    stop(false);
}



}
// no name namespace


int main(int argc, char * argv[])
{
    try
    {
        snapcluster::pointer_t cluster = std::make_shared<snapcluster>(argc, argv);
        cluster->run();
        return 0;
    }
    catch(std::exception const & e)
    {
        // clean error on exception
        std::cerr << "snapsignal: exception: " << e.what() << std::endl;
        return 1;
    }
}


// vim: ts=4 sw=4 et
